torch_geometric/data/feature_store.py

```diff
@@ -0,0 +1,354 @@
+        return self
+
+    def __call__(self) -> FeatureTensorType:
+        r"""Supports :obj:`AttrView` as a callable to force retrieval from
+        the currently specified attributes. In particular, this passes the
+        current :obj:`TensorAttr` object to a GET call, regardless of whether
+        all attributes have been specified. It returns the result of this
+        call. In particular, the following operation returns a Tensor by
+        performing a GET operation on the backing feature store:
+
+        .. code-block:: python
+
+            store[group_name, attr_name]()
+
+        """
+        return self._store.get_tensor(self._attr)
+
+    def __eq__(self, __o: object) -> bool:
+        r"""Compares two :obj:`AttrView` objects by checking equality of their
+        :obj:`FeatureStore` references and :obj:`TensorAttr` attributes."""
+        if not isinstance(__o, AttrView):
+            return False
+
+        return id(self._store) == id(__o._store) and self._attr == __o._attr
+
+    def __repr__(self) -> str:
+        return (f'{self.__class__.__name__}(store={self._store}, '
+                f'attr={self._attr})')
+
+
+class FeatureStore(MutableMapping):
+    def __init__(self, attr_cls: Any = TensorAttr):
+        r"""Initializes the feature store. Implementor classes can customize
+        the ordering and require nature of their :obj:`TensorAttr` tensor
+        attributes by subclassing :class:`TensorAttr` and passing the subclass
+        as `attr_cls`."""
+        super().__init__()
+        self._attr_cls = attr_cls
+
+    # Core (CRUD) #############################################################
+
+    @abstractmethod
+    def _put_tensor(self, tensor: FeatureTensorType, attr: TensorAttr) -> bool:
+        r"""Implemented by :obj:`FeatureStore` subclasses."""
+        pass
+
+    def put_tensor(self, tensor: FeatureTensorType, attr: TensorAttr) -> bool:
+        r"""Synchronously adds a :obj:`FeatureTensorType` object to the feature
+        store.
+
+        Args:
+            tensor (FeatureTensorType): the features to be added.
+            attr (TensorAttr): any relevant tensor attributes that correspond
+                to the feature tensor. See the :obj:`TensorAttr` documentation
+                for required and optional attributes. It is the job of
+                implementations of a FeatureStore to store this metadata in a
+                meaningful way that allows for tensor retrieval from a
+                :obj:`TensorAttr` object.
+        Returns:
+            bool: whether insertion was successful.
+        """
+        attr = self._attr_cls.cast(attr)
+        return self._put_tensor(tensor, attr)
+
+    @abstractmethod
+    def _get_tensor(self, attr: TensorAttr) -> Optional[FeatureTensorType]:
+        r"""Implemented by :obj:`FeatureStore` subclasses."""
+        pass
+
+    def get_tensor(self, attr: TensorAttr) -> Optional[FeatureTensorType]:
```

Comment:
```suggestion
    def get_tensor(self, *args, **kwargs) -> FeatureTensorType:
```
see above. In addition, I don't understand the `Optional[...]` output.

Reply:
The Optional output is to indicate that `get` may return `None`, if the tensor is not found.

Reply:
Python dictionaries throw an error in that case. Do we want to break with this?

Reply:
Fair enough. Since it doesn't really ever make sense for a feature store to store `None`, we can raise the `KeyError` in `get`.